pyccel/codegen/printing/cwrappercode.py

```diff
@@ -51,22 +52,47 @@ def get_cast_function(self, used_names, cast_type, from_variable, to_variable):
         if cast_type in self._cast_functions_dict:
             return self._cast_functions_dict[cast_type]
 
+        cast_function_name = self.get_new_name(used_names, cast_type)
         cast_function_arg = [from_variable]
         cast_function_result = [to_variable]
-        cast_function_ret = Return(cast_function_result)
-        cast_function_name = self.get_new_name(used_names, cast_type)
-        cast_function = CastFunction(cast_function_name, cast_type,
-                            cast_function_arg, cast_function_ret, cast_function_result)
+        cast_function_local_vars = []
+ 
+        #switch case of cast_type
+        if cast_type == 'pyint_to_bool':
+            cast_function_body = [Assign(cast_function_result[0], Bool(cast_function_arg[0]))]
+        elif cast_type == 'bool_to_pyobj':
+            cast_function_body = [If((Bool(cast_function_arg[0]),
+                [Assign(cast_function_result[0], Symbol('Py_True'))]),
+                (BooleanTrue(), [Assign(cast_function_result[0], Symbol('Py_False'))]))]
+        elif cast_type == 'pycomplex_to_complex':
+            cast_function_body = [Assign(cast_function_result[0],
+                FuncCall('__builtin_complex', [FuncCall('PyComplex_RealAsDouble', cast_function_arg),
+                                               FuncCall('PyComplex_ImagAsDouble', cast_function_arg)]))]
+        elif cast_type == 'complex_to_pycomplex':
+            cast_function_body = [Assign(cast_function_result[0],
+                FuncCall('PyComplex_FromDoubles', [FuncCall('__real', cast_function_arg),
+                                                   FuncCall('__imag', cast_function_arg)]))]
+        cast_function_body += [Return(cast_function_result)]
+        cast_function = FunctionDef(name = cast_function_name,
+                                    arguments = cast_function_arg,
+                                    body = cast_function_body,
+                                    results = cast_function_result,
+                                    local_vars = cast_function_local_vars)
+        self._cast_functions_dict[cast_type] = cast_function
         return cast_function
 
     def get_PyArgParseType(self, used_names, variable):
         if variable.dtype is NativeBool():
             collect_type = NativeInteger()
             collect_var = Variable(dtype=collect_type, precision=4,
                 name = self.get_new_name(used_names, variable.name+"_tmp"))
-
             cast_function = self.get_cast_function(used_names, 'pyint_to_bool', variable, collect_var)
-
+            return collect_var , cast_function
+        if variable.dtype is NativeComplex():
+            collect_type = PyccelPyObject()
+            collect_var = Variable(dtype=collect_type, rank = 1,
```

Comment:
Why is this an array? Shouldn't it just be a pointer?

Reply:
I used this one until I found out that the printer of function_call use the printer of the variable to print arguments.
So what happen when we pass a variable  'X'  pointer as arg it will be printed like this 
```
some_function((*X))
```

Reply:
Ah, we need to add a test that compares the function definition arguments to the provided arguments to see if we should be passing a pointer or a value. I think this is already done for the return arguments.
When we have a problem like this we should fix the problem rather than working round it by pretending that we have an array instead of a pointer. Pointers have not been used in C up till now so any problems are just because the syntax hasn't been fully implemented

Reply:
I didn't know that before I just used rank = 1 believing it mean pointer.

Reply:
Ah ok. The `rank` is the number of dimensions. So a matrix (2d-array) has rank 2, a vector (1d-array) has rank 1, and a scalar (0d-array) has rank 0. In C arrays are represented using pointers so the distinction is currently not very clear in our code. But if we were writing C++ and using the `vector` structure instead of C-style arrays then the declaration of `Variable(dtype='int', 'a', rank = 1)` would be `std::vector<int> a` whereas the declaration of `Variable(dtype='int', 'a', is_pointer = True)` would be `int* a`